packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -187,9 +188,25 @@ export class ModularChangeFamily
 			return makeModularChangeset();
 		}
 
-		return changes
-			.map((change) => change.change)
-			.reduce((change1, change2) => this.composePair(change1, change2, revInfos, idState));
+		const innerChanges = changes.map((change) => change.change);
+
+		// Recursively compose the left and right halves of the changeset list before composing their respective compositions.
+		// This leads to the same number of compositions as when composing the changesets in order,
+		// but in scenarios where the changesets grow as they are composed
+		// (e.g., when composing changesets that target different elements of a sequence),
+		// this approach reduces the number of times the individual change atoms within the changesets are processed.
+		// It benefits from the same principle that makes merge sort faster than insertion sort,
+		// making it O(N*log(N)) instead of O(N²) for N changesets each containing 1 change atom.
+		const balancedCompose = (slice: readonly ModularChangeset[]): ModularChangeset => {
```

Comment:
if you want, you could factor this out function to be able to apply to any monadic operation as a higher order function (maybe balancedReduce over any array, maybe put it in utils). This would allow it to be trivially unit tested with string composition.

Reply:
I think this implementation doesn't handle the empty case. Having it factored out would make testing that kind of thing much easier. I'm not sure how we want to handle that (assert, debugAssert, throw etc), but documenting and testing it might be good

Reply:
Handling the empty case the same way as array.reduce would be nice: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce#edge_cases

(move the empty check and use of initial value into this function so its total and no longer has an error case). Maybe make the initalValue a callback though so we don't compute it when unneeded, and maybe rename it "emptyCase" if we don't use it in other cases.